ast/compile.go

```diff
@@ -844,12 +858,114 @@ func (c *Compiler) checkSafetyRuleHeads() {
 	}
 }
 
+func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+	var refLoader gojsonschema.JSONLoader
+	sl := gojsonschema.NewSchemaLoader()
+
+	if goSchema != nil {
+		refLoader = gojsonschema.NewGoLoader(goSchema)
+	} else {
+		return nil, fmt.Errorf("no schema as input to compile")
+	}
+	schemasCompiled, err := sl.Compile(refLoader)
+	if err != nil {
+		return nil, fmt.Errorf("unable to compile the schema due to: %s", err.Error())
+	}
+	return schemasCompiled, nil
+}
+
+func parseSchema(schema interface{}) (types.Type, error) {
+	subSchema, ok := schema.(*gojsonschema.SubSchema)
+	if !ok {
+		return nil, fmt.Errorf("unexpected schema type %v", subSchema)
+	}
+
+	// Handle referenced schemas, returns directly when a $ref is found
+	if subSchema.RefSchema != nil {
+		return parseSchema(subSchema.RefSchema)
+	}
+
+	if subSchema.Types.IsTyped() && subSchema.Types.Contains(`boolean`) {
```

Comment:
💭 Whenever I see strings with backticks, I wonder what's special about them. Here, nothing is, so I think we could also go with `"boolen"`, `"string"` etc. Anyways, it's just nitpicking. Don't bother 😄

Reply:
Done (commit https://github.com/open-policy-agent/opa/pull/3060/commits/2616431346afb485dd41e56d2b5f68100fb41d87)